server/coordinator/server.go

```diff
@@ -0,0 +1,335 @@
+package coordinator
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/privacylab/talek/bloom"
+	"github.com/privacylab/talek/common"
+	"github.com/privacylab/talek/cuckoo"
+	"golang.org/x/net/trace"
+)
+
+// Server is the main logic for the central coordinator
+type Server struct {
+	/** Private State **/
+	// Static
+	log               *common.Logger
+	name              string
+	snapshotThreshold uint64
+	snapshotInterval  time.Duration
+
+	// Thread-safe (locked)
+	lock          *sync.Mutex
+	config        common.Config // Config
+	servers       []NotifyInterface
+	commitLog     []*CommitArgs // Append and read only
+	numNewCommits uint64
+	snapshotCount uint64
+	lastLayout    []uint64
+	intVec        []uint64
+	cuckooData    []byte
+	cuckooTable   *cuckoo.Table
+
+	// Channels
+	notifyChan chan bool
+	closeChan  chan bool
+}
+
+// NewServer creates a new Centralized talek server.
+func NewServer(name string, config common.Config, servers []NotifyInterface, snapshotThreshold uint64, snapshotInterval time.Duration) (*Server, error) {
+	s := &Server{}
+	s.log = common.NewLogger(name)
+	s.name = name
+	s.snapshotThreshold = snapshotThreshold
+	s.snapshotInterval = snapshotInterval
+
+	s.lock = &sync.Mutex{}
+	s.config = config
+	if servers == nil {
+		s.servers = make([]NotifyInterface, 0)
+	} else {
+		s.servers = servers
+	}
+	s.commitLog = make([]*CommitArgs, 0)
+	s.numNewCommits = 0
+	s.snapshotCount = 0
+	s.lastLayout = make([]uint64, config.NumBuckets*config.BucketDepth)
+	s.intVec = buildInterestVector(config.WindowSize(), config.BloomFalsePositive, s.commitLog[:]).Bytes()
+	s.cuckooData = make([]byte, config.NumBuckets*config.BucketDepth*uint64(common.IDSize))
+
+	// Choose a random seed for the cuckoo table
+	seedBytes := make([]byte, 8)
+	_, err := rand.Read(seedBytes)
+	if err != nil {
+		s.log.Error.Printf("coordinator.NewServer(%v) error: %v", name, err)
+		return nil, err
+	}
+	seed, _ := binary.Varint(seedBytes)
+	s.cuckooTable = cuckoo.NewTable(name, config.NumBuckets, config.BucketDepth, uint64(common.IDSize), s.cuckooData, seed)
+	// Should not be possible
+	if s.cuckooTable == nil {
+		err := fmt.Errorf("Invalid cuckoo table parameters")
+		s.log.Error.Printf("coordinator.NewServer(%v) error: %v", name, err)
+		return nil, err
+	}
+	s.notifyChan = make(chan bool)
+	s.closeChan = make(chan bool)
+
+	go s.loop()
+
+	s.log.Info.Printf("coordinator.NewServer(%v) success\n", name)
+	return s, nil
+}
+
+/**********************************
+ * PUBLIC RPC METHODS (threadsafe)
+ **********************************/
+
+// GetInfo returns information about this server
+func (s *Server) GetInfo(args *interface{}, reply *GetInfoReply) error {
+	tr := trace.New("Coordinator", "GetInfo")
+	defer tr.Finish()
+	s.lock.Lock()
```

Comment:
is it worth having an RWMutex, to support multiple-reader on all but the commit method?